File: tests/test_runtime_task.py
import subprocess
import tempfile
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]


def test_every_runtime_task_is_created_from_its_pinned_schedule_entry():
    with tempfile.TemporaryDirectory() as tmp:
        tmpdir = Path(tmp)
        (tmpdir / "freertos").mkdir()
        (tmpdir / "esp_err.h").write_text(
            "#pragma once\ntypedef int esp_err_t;\n"
            "#define ESP_OK 0\n#define ESP_ERR_INVALID_ARG 0x102\n#define ESP_ERR_NO_MEM 0x101\n"
        )
        (tmpdir / "esp_log.h").write_text(
            "#pragma once\nvoid test_log(const char *, const char *, ...);\n"
            "#define ESP_LOGE(...) test_log(__VA_ARGS__)\n"
        )
        (tmpdir / "freertos" / "FreeRTOS.h").write_text(
            "#pragma once\n#include <stdint.h>\n"
            "typedef int BaseType_t; typedef unsigned UBaseType_t;\n"
            "#define pdPASS 1\n"
        )
        (tmpdir / "freertos" / "task.h").write_text(
            "#pragma once\n#include <stdint.h>\n#include \"freertos/FreeRTOS.h\"\n"
            "typedef void *TaskHandle_t; typedef void (*TaskFunction_t)(void *);\n"
            "BaseType_t xTaskCreatePinnedToCore(TaskFunction_t, const char *, uint32_t, "
            "void *, UBaseType_t, TaskHandle_t *, BaseType_t);\n"
        )
        binary = tmpdir / "runtime_task_test"
        subprocess.run(
            [
                "cc", "-std=c11", "-Wall", "-Wextra", "-Werror", "-DESP_PLATFORM",
                "-I", str(tmpdir), "-I", str(ROOT / "main"),
                str(ROOT / "main" / "runtime_schedule.c"),
                str(ROOT / "main" / "runtime_task.c"),
                str(ROOT / "tests" / "test_runtime_task.c"),
                "-o", str(binary),
            ],
            check=True,
        )
        subprocess.run([str(binary)], check=True)

File: tests/test_runtime_task.c
#include <assert.h>
#include <stddef.h>
#include <stdint.h>
#include <string.h>

#include "runtime_schedule.h"
#include "runtime_task.h"

static runtime_task_id_t registered_id;
static TaskHandle_t registered_handle;

BaseType_t xTaskCreatePinnedToCore(TaskFunction_t fn, const char *name,
                                   uint32_t stack_size, void *arg,
                                   UBaseType_t priority, TaskHandle_t *out,
                                   BaseType_t core)
{
    (void)arg;
    assert(fn != NULL);
    const runtime_task_spec_t *spec = runtime_schedule_get(registered_id);
    assert(spec != NULL);
    assert(strcmp(name, spec->name) == 0);
    assert(stack_size == spec->stack_size);
    assert(priority == spec->priority);
    assert(core == spec->core);
    *out = (TaskHandle_t)(uintptr_t)(registered_id + 1);
    registered_handle = *out;
    return pdPASS;
}

void runtime_metrics_register_handle(runtime_task_id_t id, TaskHandle_t handle)
{
    assert(id == registered_id);
    assert(handle == registered_handle);
}

void test_log(const char *tag, const char *format, ...)
{
    (void)tag;
    (void)format;
}

static void test_task(void *arg)
{
    (void)arg;
}

int main(void)
{
    assert(runtime_schedule_validate());
    for (registered_id = RUNTIME_TASK_CONTROL;
         registered_id < RUNTIME_TASK_COUNT;
         ++registered_id) {
        registered_handle = NULL;
        TaskHandle_t out = NULL;
        assert(runtime_task_create(registered_id, test_task, NULL, &out) == ESP_OK);
        assert(out == (TaskHandle_t)(uintptr_t)(registered_id + 1));
    }
    return 0;
}
